src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+            // If the message has been fully sent, move on to the next one.
+            if (self.bytes_sent == self.send_queue.out.?.message.buffer.len) {
+                self.bytes_sent = 0;
+                const envelope = self.send_queue.pop().?;
+                envelope.message.references -= 1;
+                self.message_bus.gc(envelope.message);
+                self.message_bus.allocator.destroy(envelope);
+            }
+            self.send();
+        }
+    };
+
+    /// Connection used to recive data from clients and replicas as well as
+    /// send data to clients. This connection is not re-created if it drops,
+    /// The other replica or client is responsible for re-connecting.
+    const Connection = struct {
+        message_bus: *MessageBus,
+        /// A value of -1 indicates that the connection is currently being
+        /// closed or is inactive.
+        fd: os.socket_t = -1,
+        /// The peer is determined by inspecting the first message header
+        /// received.
+        peer: union(enum) {
+            /// No peer is currently connected.
+            none: void,
+            /// A connection has been established but the first header has not yet been recieved.
+            unknown: void,
+            /// The peer is a client with the given id.
+            client: u128,
+            /// The peer is a replica with the given id.
+            replica: u16,
+        } = .none,
+
+        /// This completion is used for all recv operations.
+        recv_completion: IO.Completion = undefined,
+        /// Number of bytes of the current header/message that have already been recieved.
+        bytes_received: usize = 0,
+        incoming_header: Header = undefined,
+        incoming_message: *Message = undefined,
+
+        /// This completion is used for all send operations.
+        send_completion: IO.Completion = undefined,
+        /// The queue of messages to send to the client.
+        /// Empty unless peer == .client
+        send_queue: FIFO(Envelope) = .{},
+        bytes_sent: usize = 0,
+
+        /// Given a newly accepted fd, start receiving messages on it.
+        /// Callbacks will be continously re-registered until close() is
+        /// called and the connection is terminated.
+        fn receive_messages(self: *Connection, fd: os.socket_t) void {
+            assert(self.fd == -1);
+            assert(self.peer == .none);
+            self.fd = fd;
+            self.peer = unknown;
+            self.recv_header();
+        }
+
+        fn recv_header(self: *Connection) void {
+            // The fd may have been closed due to an error while sending.
+            if (self.fd == -1) return;
+            self.message_bus.io.recv(
+                *ReplicaConnection,
+                self,
+                complete_recv_header,
+                &self.completion,
+                mem.asBytes(&self.incoming_header)[self.bytes_received..],
+                os.MSG_NOSIGNAL,
```

Comment:
Good call on your part to pass `os.MSG_NOSIGNAL` explicitly rather than making it the default in IO.